crowbar_framework/lib/upgrade.rb

```diff
@@ -0,0 +1,71 @@
+class Upgrade
+  attr_accessor :upgrade_progress
+
+  # Return the current state of upgrade process.
+  # We're keeping the information in the file so is accessible by
+  # external applications and different crowbar versions.
+  def initialize
+    @upgrade_progress = {
+      current_step: upgrade_steps_6_7.first,
+      # substep is needed for more complex steps like upgrading the nodes
+      current_substep: nil,
+      # current node is relevant only for the nodes_upgrade step
+      current_node: nil
+    }
+    if progress_file_path.exist?
+      @upgrade_progress = JSON.load(progress_file_path.read)
+    else
+      # in 'steps', we save the information about each step that was executed
+      @upgrade_progress[:steps] = upgrade_steps_6_7.map do |step|
+        [step, { status: "pending", errors: {} }]
+      end.to_h
+    end
+  end
+
+  def save
+    progress_file_path.open("w") do |f|
+      f.write(JSON.pretty_generate(upgrade_progress))
+    end
+  end
+
+  def current_substep
+    upgrade_progress[:current_substep]
+  end
+
+  def current_step
+    upgrade_progress[:current_step]
+  end
+
+  def next_step!
+    return true if finished?
```

Comment:
maybe return false when finished? or how shall the caller determin the end when this function is used a few times?